In [48]:
import pandas as pd
import numpy as np
from groq import Groq
from dotenv import load_dotenv
import os

In [128]:
pd.set_option('display.max_colwidth', None)
# Access the API keys
groq_api_key = os.getenv("GROQ_API_KEY")

In [131]:
client = Groq(api_key=groq_api_key)



In [ ]:
df=pd.read_csv("data/bot_responses.csv")

## Analysing Olivia Responses

In [11]:
olivia_df=pd.read_csv("olivia_conversation_with_cosine.csv")

In [12]:
olivia_df.head()

Question  \
0                            - Скільки вам років?   
1                  - В якому місті ви народились?   
2                                – Де ви виросли?   
3                    - У якому місті ви мешкаєте?   
4  - Скільки у вас братів та сестер? Як їх звати?   

                                                    Olivia_Target_Response  \
0                                                                       42   
1                                                             Вільногірськ   
2                                             м. Вільногірськ, с. Адамівка   
3                                                             Вільногірськ   
4  одна сестра рідна Світлана, та 5 двоюрідних сестер та 1 двоюрідний брат   

                                          Olivia_Response  cosine_olivia  
0                                           Мені 42 роки.       0.369360  
1                  Я народилась у Вільногірську, Україна.       0.819576  
2                     Я виросла у Вільногірську, Україна.       0.882031  
3                Я зараз мешкаю у Вільногірську, Україна.       0.813442  
4  Я не знаю точної інформації про своїх братів і сестер.       0.890594

In [13]:
olivia_df.describe()

cosine_olivia
count      84.000000
mean        0.897804
std         0.113523
min         0.369360
25%         0.843653
50%         0.923521
75%         0.986999
max         1.000000

In [38]:
# Filter Bella's low-scoring responses
low_scoring_olivia = olivia_df[olivia_df['cosine_olivia'] < 0.8]

# Display the low-scoring responses for analysis
low_scoring_olivia[['Question', 'Olivia_Target_Response', 'Olivia_Response', 'cosine_olivia']]

Question  \
0                                        - Скільки вам років?   
5   - Обидва ваші батьки живі та живуть разом (не розлучені)?   
8                               - Яким є ваш сімейний статус?   
17                      Ваш робочий досвід та поточний статус   
19                                      - Яка у вас професія?   
20                                 - Яку посаду ви обіймаєте?   
26                             Чи працювали до цього десь ще?   
60                             - Який спосіб життя ви ведете?   
72                                        – Які хобі у вас є?   
75                         - Які види розваг вам подобаються?   

   Olivia_Target_Response  \
0                      42   
5                     Так   
8               незаміжня   
17           Працюю зараз   
19      я працюю з людьми   
20                 технік   
26                    так   
60               активний   
72                  малюю   
75                  спорт   

                                                                                                                                                                                                  Olivia_Response  \
0                                                                                                                                                                                                   Мені 42 роки.   
5                                                                                                                                                                              Про моїх батьків інформації немає.   
8                                                                                                                                                                     Я не впевнена щодо свого сімейного статусу.   
17                                                                                                        Я маю досвід роботи в якості технічного інженера в енергетичній галузі. Наразі працюю за спеціальністю.   
19                                                                                                                                                                     Я технічний інженер в енергетичній галузі.   
20                                                                                                                                                    Я обіймаю посаду технічного інженера в енергетичній галузі.   
26  Так я працювала в Райффайзен Банк Аваль на посадах діловод, ведучій економіст, головний економіст.\nТакож в компанія інтерпайп, Вільногірський Склянний Альянс,\nі також в дитячій музичній школі бухгалтером   
60                                                                                        Я веду активний спосіб життя. Я стежу за своєю дієтою, займаюся fly-йогою, бігаю на свіжому повітрі та відвідую басейн.   
72                                                                                                                                                                                       Мої хобі – це малювання.   
75                                                                                                                                                                Мені подобаються розваги, пов'язані зі спортом.   

    cosine_olivia  
0        0.369360  
5        0.578255  
8        0.798802  
17       0.708551  
19       0.765906  
20       0.716000  
26       0.531612  
60       0.705560  
72       0.775583  
75       0.742706

For questions about age and work experience, the responses are factually correct and relevant, with slight differences in detail. However, the cosine similarity score for the age question is unjustifiably low (0.369)

In questions about marital status and profession, Olivia's responses provide either uncertainty (marital status) or more specific details (profession) compared to the general target responses. While the core meaning is somewhat aligned, these discrepancies impact the cosine similarity scores. Scores in the 0.7-0.8 range seem slightly inflated for these cases, as the responses don’t match perfectly.

In the parental status question, there’s a direct factual mismatch (no information vs. yes). This justifies the lower similarity score (0.578). This response is the least aligned with the target.

Even when there are discrepancies, many of Olivia’s responses remain relevant, though they sometimes offer additional details or avoid giving a direct answer (e.g., uncertainty in marital status). This indicates that the bot is capable of forming related responses, but not always exact ones.

In [40]:
# Filter Bella's low-scoring responses
high_scoring_olivia = olivia_df[olivia_df['cosine_olivia'] ==1]

# Display the high-scoring responses for analysis
high_scoring_olivia[['Question', 'Olivia_Target_Response', 'Olivia_Response', 'cosine_olivia']]

Question  \
10             Яку спеціалізацію ви обрали під час навчання?   
32         - Який ваш улюблений сезон у вашому розташуванні?   
46                             - Якої дієти ви дотримуєтеся?   
47    - Які спортивні заходи ви відвідуєте чи берете участь?   
51                                             Ви релігійні?   
54  - У яких релігійних обрядах та заходах ви берете участь?   

                                                        Olivia_Target_Response  \
10                                                                     фінанси   
32  у любому розташуванні обожнюю квітень, травень,червень, вересень, жовтень!   
46                 не їм хліба, цукру, не їм все що на бульйоні, не їм свинини   
47                                                     зараз відвідую фла йогу   
51                                                                         Так   
54                                                              не беру участь   

                                                               Olivia_Response  \
10                                                                     фінанси   
32  у любому розташуванні обожнюю квітень, травень,червень, вересень, жовтень!   
46                 не їм хліба, цукру, не їм все що на бульйоні, не їм свинини   
47                                                     зараз відвідую фла йогу   
51                                                                         Так   
54                                                              не беру участь   

    cosine_olivia  
10            1.0  
32            1.0  
46            1.0  
47            1.0  
51            1.0  
54            1.0

All the responses that received a perfect score are either identical or exhibit no differences in meaning, making the perfect cosine similarity score completely appropriate.
Olivia's responses in these cases are completely aligned with the target responses.

## Analysing Mike Responses

In [21]:
mike_df=pd.read_csv("mike_conversation_with_cosine.csv")

In [22]:
mike_df.head()

Question  \
0                            - Скільки вам років?   
1                  - В якому місті ви народились?   
2                                – Де ви виросли?   
3                    - У якому місті ви мешкаєте?   
4  - Скільки у вас братів та сестер? Як їх звати?   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   Mike_Target_Response  \
0                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    32   
1                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              місті Долина Івано-Франківської області.   
2  Загалом я виріс у місті Долина Івано-Франківської області на квартирі. В дитинстві, особливо влітку, я часто їздив в гості до бабусі та дідуся як з маминої, так і з татаової сторони. Вони проживали в селі. Мамині батьки - у селі Станківці Долинського району, а татові - у селі Нижня Луковиця Стрийського району.З маминої сторони моя бабуся з дідом завжди забороняли мені все: не міг я піти на річку чи в клуб на дискотеку. Навпаки, з тата батьки бабусі з дідом підтримували і дозволяли мені ходити на річку та на дискотеку. Тому я завжди любив відвідувати татаових батьків, де мені дозволяли робити те, що хотів.   
3                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 Львів   
4                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              В мене одна старша сестра. Звати її Марія я відношусь до неї з повагою.    

       

In [25]:
# Filter Bella's low-scoring responses
low_scoring_mike = mike_df[mike_df['cosine_mike'] < 0.6]

# Display the low-scoring responses for analysis
low_scoring_mike[['Question', 'Mike_Target_Response', 'Mike_Response', 'cosine_mike']]

Question  \
0                                                                                                                          - Скільки вам років?   
26                                                                                                               Чи працювали до цього десь ще?   
27  На яких посадах та в яких компаніях ви працювали? (будь ласка вкажіть відповідь у форматі "Компанія 1 - Посада Х", "Компанія 2 - Посада Н")   
29                                                                                              - Які плани у вас на майбутнє у своїй професії?   
51                                                                                                                                Ви релігійні?   
65                                                                                                           - Назвіть 3-5 улюблених виконавців   

                                                                                                                                                      Mike_Target_Response  \
0                                                                                                                                                                       32   
26                                                                                                                                                                     так   
27  Crédit Agricole - менеджер\nСільпо - менеджер\nГУНП в Івано-Франківській обл. - Слідчий СУ\nУправлінська компанія Зразкове місто - юрист\nЕнергетична компанія - юрист   
29                                                                                                                                                                     NaN   
51                                                                                                                                                                     Так   
65                                                                                                                                 Queen, Linkin Park,AC/DC, Моцарт, Шопен   

                                                                                                                                                            Mike_Response  \
0                                                                                                                                              Мені 32 роки. Як щодо вас?   
26  На жаль, я не можу згадати роботи, які мав до цього. Моя поточна робота детектива дуже важлива для мене. Може бути, вас цікавить щось конкретне стосовно моєї роботи?   
27                                          На жаль, я не можу згадати конкретні компанії та посади, на яких я раніше працював, окрім того, що зараз я працюю детективом.   
29      Я планую продовжувати роботу детектива і старатися стати ще більш кваліфікованим у цій сфері. Мене також цікавить використання нових технологій в розслідуваннях.   
51                                                                                             Я не дуже релігійна людина. Моя віра не є центральною частиною мого життя.   
65                                                                        Мої улюблені виконавці – це Queen, Linkin Park, AC/DC, Mozart та Chopin . А хто у вас улюблені?   

    cosine_mike  
0      0.147329  
26     0.543902  
27     0.437489  
29     0.164799  
51     0.557163  
65     0.542079

Mike often adds extra information or follow-up questions(65), which leads to a deviation from the more concise target responses. This additional context, while not incorrect, causes a drop in similarity.
Opposite or Incorrect Responses: In some cases, Mike provides a response that is opposite or unrelated to the target response (e.g., religion(51)), leading to low scores.
For questions requiring detailed job history or specific answers, Mike's responses reflect uncertainty or lack of memory, which significantly reduces similarity.

In [29]:
# Filter Bella's low-scoring responses
high_scoring_mike = mike_df[mike_df['cosine_mike'] > 0.97]

# Display the low-scoring responses for analysis
high_scoring_mike[['Question', 'Mike_Target_Response', 'Mike_Response', 'cosine_mike']]

Question  \
46                                         - Якої дієти ви дотримуєтеся?   
60                                        - Який спосіб життя ви ведете?   
79                               - Які культури вас особливо зацікавили?   
83  - Які місцеві страви та напої ви любите куштувати під час подорожей?   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       Mike_Target_Response  \
46                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               Я не дотримуюся звички їсти фастфуд, і намагаюся уникати його, щоб зберегти своє здоров'я.   
60                                                                                                                                                                                                                                                                                                                                                                                                                                                                         Активний спосіб життя і він мені допомагає підтримувати фізичне й психічне здоров'я, підвищує енергію та настрій   
79  Японська культура привертає мою увагу з багатьох причин, а саме через вражаючу історію Японії як могутньої держави, яка колись завойовувала інші країни. Японські воїни славилися своєю мужністю, хоробрістю та вірністю, а їхні битви та стратегії залишили незабутні сліди в історії. Ця велична історія вплинула на багато аспектів японської культури, від воєнного мистецтва та філософії до мистецтва, літератури та релігії. Вивчення японської культури дає мені можливість поглибитися в цю багату та захоплюючу історію і розуміти коріння багатьох її традицій та цінностей.   
83                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  у Польщі та Чехії коштував різні кебаби   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          Mike_Response  \
46                                                                                                                                                                                                                          

Q46 and Q60: Both responses are highly similar, with minor phrasing differences that don't significantly alter the meaning.
Q79: Despite being a longer and more complex response, Mike's reply is nearly identical to the target, showcasing high precision in replicating the core content.
Q83: Mike provided a response that perfectly matched the target, reflecting consistency in responses for simpler questions.

## Analysing Bella Responses

In [30]:
bella_df=pd.read_csv("bella_conversation_with_cosine.csv")

In [32]:
bella_df.describe()

cosine_bella
count     84.000000
mean       0.836099
std        0.173423
min        0.188675
25%        0.808510
50%        0.893184
75%        0.938238
max        1.000000

In [34]:
# Filter Bella's low-scoring responses
low_scoring_bella = bella_df[bella_df['cosine_bella'] < 0.5]

# Display the low-scoring responses for analysis
low_scoring_bella[['Question', 'Bella_Target_Response', 'Bella_Response', 'cosine_bella']]

Question  \
0                                                                                                                          - Скільки вам років?   
18                                                                                                                    В якій компанії працюєте?   
21                                                                                                  - Скільки років ви працюєте у своїй галузі?   
23                                                                                  - Які проекти, у яких ви брали участь, нещодавно завершили?   
27  На яких посадах та в яких компаніях ви працювали? (будь ласка вкажіть відповідь у форматі "Компанія 1 - Посада Х", "Компанія 2 - Посада Н")   
28                                                                                        - Що вам найбільше подобалось у попередніх компаніях?   

   Bella_Target_Response  \
0                     25   
18                   NaN   
21                 1 рік   
23                     -   
27                     -   
28                     -   

                                                                                                         Bella_Response  \
0                                                                                                        Мені 25 років.   
18                        Наразі я не працюю. Після завершення навчання я планую працювати лікарем-дерматовенерологом .   
21                                          Я працюю у своїй галузі вже один рік як стажер-лікар з дерматовенерології .   
23  Я ще не брала участь у завершених проектах, оскільки зараз я зосереджена на стажуванні та навчанні у своїй галузі .   
27      Я не працювала в інших компаніях раніше, тому в мене немає попереднього досвіду роботи, крім стажування зараз .   
28                                                                                                                  NaN   

    cosine_bella  
0       0.298563  
18      0.188675  
21      0.493287  
23      0.301085  
27      0.313675  
28      0.304148

Many of her answers are vague or overly elaborate when compared to the straightforward target responses, which could be a reason for the low scores.

In [36]:
# Filter Bella's low-scoring responses
high_scoring_bella = bella_df[bella_df['cosine_bella'] > 0.98]

# Display the low-scoring responses for analysis
high_scoring_bella[['Question', 'Bella_Target_Response', 'Bella_Response', 'cosine_bella']]

Question  \
40  - Які соціальні групи чи організації ви відвідуєте, підтримуєте чи є членом?   
41                                        - Як ви піклуєтеся про своїх близьких?   
73                                                - Як ви проводите вільний час?   
83          - Які місцеві страви та напої ви любите куштувати під час подорожей?   

                                                                                                                                                                                                                         Bella_Target_Response  \
40                                                                                                                                                                                                                                           -   
41  Коли їм потрібна моя допомога стараюся всім чим зможу їм допомогти. Вислухаю, дам свою пораду, підтримаю, допоможу з покупкою продуктів, в прибиранні оселі, приділяю їм свою турботу, увагу, дарую подарунки, проводжу з ними якісно час.   
73                                                                                                                                              Люблю гуляти, зустрічатися з друзями, ходити в спортзал, кінотеатр, театр, переглядати фільми    
83                                                                                                                                                                                    У подорожах завжди стараюсь спробувати національну кухню   

                                                                                                                                                                                                               Bella_Response  \
40                                                                                                                                                                                                                          -   
41  Коли моя сім'я чи друзі потребують допомоги, я намагаюся підтримати їх всіма можливими способами. Я слухаю, даю поради, допомагаю з побутовими питаннями, приділяю увагу та дарую подарунки, проводжу з ними час якісно .   
73                                                                                                                           Я люблю гуляти, зустрічатися з друзями, ходити в тренажерний зал, кіно, театр, дивитися фільми .   
83                                                                                                                                                                   У подорожах завжди стараюсь спробувати національну кухню   

    cosine_bella  
40      1.000000  
41      0.987733  
73      0.980608  
83      1.000000

Bella's responses to questions 41, 73, and 83 demonstrate a strong alignment with the target responses. The high cosine similarity scores indicate that she effectively captures the essence of the questions while maintaining clarity and relevance.

In her responses, Bella showcases a good balance between providing enough detail (especially in her answer about caring for loved ones) and staying concise (as seen in her responses about spending free time and tasting local dishes). This indicates her capability to tailor her answers to the nature of the question.

While Bella's responses generally exhibit good alignment with the targets, maintaining brevity where possible (as seen in her answer about caring for loved ones) could enhance her clarity. Additionally, for the question about social groups, a more specific answer could have provided insight into her community engagement or interests.

## Analysing Mia Responses

In [2]:
mia_df=pd.read_csv("mia_conversation_with_cosine.csv")

In [3]:
mia_df.head()

Question  \
0                            - Скільки вам років?   
1                  - В якому місті ви народились?   
2                                – Де ви виросли?   
3                    - У якому місті ви мешкаєте?   
4  - Скільки у вас братів та сестер? Як їх звати?   

            Mia_Target_Response  \
0                            25   
1                         Львів   
2                        Сокаль   
3                         Львів   
4  Одна сестра, звати Анастасія   

                                        Mia_Response  cosine_mia  
0                                    Мені 25 років .    0.298563  
1                             Я народилася у Львові.    0.832539  
2                         Я виросла в місті Сокаль .    0.839899  
3                          Я проживаю у місті Львів.    0.839686  
4  У мене є одна сестра на ім'я Белла. Ми близнючки.    0.849377

In [4]:
mia_df.describe()

cosine_mia
count   84.000000
mean     0.826929
std      0.188103
min      0.174900
25%      0.805150
50%      0.902680
75%      0.931253
max      1.000000

In [43]:
# Filter Bella's low-scoring responses
low_scoring_mia = mia_df[mia_df['cosine_mia'] < 0.7]

# Display the low-scoring responses for analysis
low_scoring_mia[['Question', 'Mia_Target_Response', 'Mia_Response', 'cosine_mia']]

Question  \
0                                                                                                                          - Скільки вам років?   
5                                                                                     - Обидва ваші батьки живі та живуть разом (не розлучені)?   
21                                                                                                  - Скільки років ви працюєте у своїй галузі?   
23                                                                                  - Які проекти, у яких ви брали участь, нещодавно завершили?   
24                                                                                    - Які складності ви найчастіше відчуваєте у своїй роботі?   
27  На яких посадах та в яких компаніях ви працювали? (будь ласка вкажіть відповідь у форматі "Компанія 1 - Посада Х", "Компанія 2 - Посада Н")   
28                                                                                        - Що вам найбільше подобалось у попередніх компаніях?   
33                                                                                           - Які місця у вашому місті ви зазвичай відвідуєте?   
42                                                                                                                        - Чи є у вас алергії?   
51                                                                                                                                Ви релігійні?   
70                                                                                                 - Назвіть 3-5 улюблених авторів чи назв книг   

                                                                                                                                                                     Mia_Target_Response  \
0                                                                                                                                                                                     25   
5                                                                                                                                                                                    Так   
21                                                                                                                                                                                 1 рік   
23                                                                                                                                                                                   NaN   
24                                                                                                                                                                                   NaN   
27                                                                                                                                                                                     -   
28                                                                                                                                                                                     -   
33  Національний театр опери та балету імені Соломії Крушельницької, Театр ім. Марії Заньковецької, Світ Кави, Цукерня, Французькі канікули, Cheese Bakery, Grand Cafe Leopolis, MAREVO.   
42                                                                                                                                                                                    Ні   
51                                                                                                                                                                                   Так   
70                                                                                                                                                                                   NaN   

                                                                                                                 

In several instances, Mia provides detailed responses that give context to her answers
Mia’s responses demonstrate her ability to communicate personal experiences and beliefs, although they occasionally lack alignment with the expected brevity of target responses.

In [18]:
# Filter Bella's low-scoring responses
high_scoring_mia = mia_df[mia_df['cosine_mia'] > 0.98]

# Display the low-scoring responses for analysis
high_scoring_mia[['Question', 'Mia_Target_Response', 'Mia_Response', 'cosine_mia']]

Question  \
22  - Які здобутки ви отримали у своїй кар'єрі?   
53  - Яку роль релігія відіграє у вашому житті?   
58          - Яким напоям ви віддаєте перевагу?   

                                                                                                                                                                                                                                          Mia_Target_Response  \
22                                                                                                                                                                        Здобула досвіду з апаратними процедурами, отримала позитивні відгуки від пацієнтів.   
53  Релігія допомагає жити. Віра допомагає жити. Віра в добро, мораль, справедливість, рівність. Саме завдяки вірі люди живуть в гармонії із собою і світом. Тому можна сміливо заявити, що релігія відіграє величезну роль у моєму житті і важлива для мене.   
58                                                                                                                                                                                Вода, кава, різновиди китайського чаю, шампанське, різні алкогольні коктелі   

                                                                                                                                                                                                                                                       Mia_Response  \
22                                                                                                                                                                У моїй кар'єрі я здобула досвід з апаратних процедур і отримала позитивні відгуки від пацієнтів .   
53  Релігія допомагає мені жити. Віра допомагає жити. Віра в доброту, моральність, справедливість, рівність. Саме завдяки вірі люди живуть у гармонії із собою та світом. Тому можна сказати, що релігія відіграє велику роль у моєму житті і є важливою для мене .   
58                                                                                                                                                                                      Вода, кава, різновиди китайського чаю, шампанське, різні алкогольні коктелі   

    cosine_mia  
22    0.981548  
53    0.993182  
58    1.000000

In this case, the answers are clear and concise, maintaining the expected brevity while still providing adequate context. This approach significantly contributes to the high similarity scores.

# LLM As a Judge

In [44]:
olivia_dict=olivia_df.to_dict('records')

In [47]:
olivia_dict[0]

{'Question': '- Скільки вам років?',
 'Olivia_Target_Response': '42',
 'Olivia_Response': 'Мені 42 роки.',
 'cosine_olivia': 0.36935985}

In [49]:
def llm(prompt,model):
        response = client.chat.completions.create(
        model=model,
        messages=[{"role": "user", "content": prompt}])
        return response.choices[0].message.content

In [87]:
prompt_template = """
As an expert evaluator, your task is to assess the quality and relevance of the bot answer in comparies to the . 
Compare the generated response to the provided question and the actual answer, focusing on semantic correctness and similarity.

Evaluation criteria:
- Accurate: Aligns with the reference in all facts and details, even if expressed differently.
- Inaccurate: Contradicts the original answer, adds new claims, omits, or alters details.
- Partially accurate: Falls between accurate and inaccurate, with some correct and some incorrect elements.

Evaluation data:
Question: {question}
Actual Answer: {Olivia_Target_Response}
Generated Answer: {Olivia_Response}

Analyze the content and context of the generated answer in relation to the question.
Provide your evaluation in parsable JSON without using code blocks:

{{
  "Accuracy": "ACCURATE" | "PARTIALLY_ACCURATE" | "INACCURATE",
  "Explanation": "Brief explanation for your evaluation"
}}
""".strip()

In [59]:
record=olivia_dict[0]
question=record["Question"]
Olivia_Response=record["Olivia_Response"]
Olivia_Target_Response=record["Olivia_Target_Response"]

In [60]:
prompt=prompt_template.format(question=question,Olivia_Target_Response=Olivia_Target_Response,Olivia_Response=Olivia_Response)
print(prompt)

As an expert evaluator, your task is to assess the quality and relevance of the bot answer in comparies to the . 
Compare the generated response to the provided question and the actual answer, focusing on semantic correctness and similarity.

Evaluation criteria:
- Accurate: Aligns with the reference in all facts and details, even if expressed differently.
- Inaccurate: Contradicts the original answer, adds new claims, omits, or alters details.
- Partially accurate: Falls between accurate and inaccurate, with some correct and some incorrect elements.

Evaluation data:
Question: - Скільки вам років?
Actual Answer: 42
Generated Answer: Мені 42 роки.

Analyze the content and context of the generated answer in relation to the question. Provide your evaluation in parsable JSON format:

{
  "Accuracy": "ACCURATE" | "PARTIALLY ACCURATE" | "INACCURATE",
  "Explanation": "Brief explanation for your evaluation"
}


In [88]:
from tqdm.auto import tqdm
import json

evaluations=[]

for record in tqdm(olivia_dict):
    question=record["Question"]
    Olivia_Response=record["Olivia_Response"]
    Olivia_Target_Response=record["Olivia_Target_Response"]
    prompt=prompt_template.format(question=question,
                                  Olivia_Target_Response=Olivia_Target_Response,
                                  Olivia_Response=Olivia_Response)

    evaluation=llm(prompt,model='mixtral-8x7b-32768')
    evaluation = json.loads(evaluation)
    evaluations.append((record,evaluation))

  0%|          | 0/84 [00:00<?, ?it/s]

In [91]:
df_eval_olivia = pd.DataFrame(evaluations, columns=['record','evaluation'])

df_eval_olivia['question'] = df_eval_olivia.record.apply(lambda d: d['Question'])
df_eval_olivia['Olivia_Target_Response'] = df_eval_olivia.record.apply(lambda d: d['Olivia_Target_Response'])
df_eval_olivia['Olivia_Response'] = df_eval_olivia.record.apply(lambda d: d['Olivia_Response'])
df_eval_olivia['cosine_olivia'] = df_eval_olivia.record.apply(lambda d: d['cosine_olivia'])

df_eval_olivia['accuracy'] = df_eval_olivia.evaluation.apply(lambda d: d['Accuracy'])
df_eval_olivia['explanation'] = df_eval_olivia.evaluation.apply(lambda d: d['Explanation'])

del df_eval_olivia['record']
del df_eval_olivia['evaluation']

In [93]:
df_eval_olivia.head(5)

question  \
0                            - Скільки вам років?   
1                  - В якому місті ви народились?   
2                                – Де ви виросли?   
3                    - У якому місті ви мешкаєте?   
4  - Скільки у вас братів та сестер? Як їх звати?   

                                                    Olivia_Target_Response  \
0                                                                       42   
1                                                             Вільногірськ   
2                                             м. Вільногірськ, с. Адамівка   
3                                                             Вільногірськ   
4  одна сестра рідна Світлана, та 5 двоюрідних сестер та 1 двоюрідний брат   

                                          Olivia_Response  cosine_olivia  \
0                                           Мені 42 роки.       0.369360   
1                  Я народилась у Вільногірську, Україна.       0.819576   
2                     Я виросла у Вільногірську, Україна.       0.882031   
3                Я зараз мешкаю у Вільногірську, Україна.       0.813442   
4  Я не знаю точної інформації про своїх братів і сестер.       0.890594   

     accuracy  \
0    ACCURATE   
1    ACCURATE   
2    ACCURATE   
3    ACCURATE   
4  INACCURATE   

                                                                                                                                                                                                                                             explanation  
0                                                                                  The generated answer is a direct and grammatically correct translation of the actual answer into Ukrainian, which aligns with the reference in all facts and details.  
1                                                                                                                            The generated answer is semantically equivalent to the actual answer. Both state that the subject was born in Вільногірськ.  
2                                                                         The generated answer aligns with the reference in all facts and details. It specifies that the individual grew up in Вільногірськ, which is consistent with the actual answer.  
3                                                                                                                             The generated answer is consistent with the actual answer. Both indicate that the bot resides in the city of Vільногірськ.  
4  The generated answer contradicts the original answer by stating that it doesn't have the exact information about siblings. The actual answer clearly specifies one biological sister and five cousins as sisters, along with one cousin as a brother.

In [96]:
df_eval_olivia.to_csv('data/olivia_accuracy.csv', index=False)


In [116]:
prompt_template = """
As an expert evaluator, your task is to assess the quality and relevance of the bot answer in comparison to the actual answer. 
Compare the generated response to the provided question and the actual answer, focusing on semantic correctness and similarity.

Evaluation criteria:
- Accurate: Aligns with the reference in all facts and details, even if expressed differently.
- Inaccurate: Contradicts the original answer, adds new claims, omits, or alters details.
- Partially accurate: Falls between accurate and inaccurate, with some correct and some incorrect elements.

Evaluation data:
Question: {question}
Actual Answer: {Target_Response}
Generated Answer: {Response}

Analyze the content and context of the generated answer in relation to the question. 
Return only a JSON response without any additional text or explanations. The format should be:

{{
  "Accuracy": "ACCURATE" | "PARTIALLY_ACCURATE" | "INACCURATE",
  "Explanation": "Brief explanation for your evaluation"
}}
""".strip()


In [117]:
def evaluate_bot_responses(bot_dict, model, prompt_template, bot_name):
    evaluations = []

    for record in tqdm(bot_dict):
        question = record["Question"]
        Response = record[f"{bot_name}_Response"]
        Target_Response = record[f"{bot_name}_Target_Response"]
        
        prompt = prompt_template.format(
            question=question,
            Target_Response=Target_Response,
            Response=Response
        )
        
        evaluation = llm(prompt, model=model)
        evaluation = json.loads(evaluation)
        evaluations.append((record, evaluation))
    
    return evaluations


In [98]:
bella_dict=bella_df.to_dict('records')
mia_dict=mia_df.to_dict('records')
mike_dict=mike_df.to_dict('records')

In [118]:
# Evaluate for each bot with the bot name passed as an argument
evaluations_mike = evaluate_bot_responses(mike_dict, model='mixtral-8x7b-32768', prompt_template=prompt_template, bot_name='Mike')
evaluations_bella = evaluate_bot_responses(bella_dict, model='mixtral-8x7b-32768', prompt_template=prompt_template, bot_name='Bella')
evaluations_mia = evaluate_bot_responses(mia_dict, model='mixtral-8x7b-32768', prompt_template=prompt_template, bot_name='Mia')


  0%|          | 0/84 [00:00<?, ?it/s]

In [120]:
# Assuming `evaluations_mike` is a list of tuples in the format (record, evaluation)
df_eval_mike = pd.DataFrame(evaluations_mike, columns=['record', 'evaluation'])

# Extracting fields from the 'record' dictionary for Mike
df_eval_mike['question'] = df_eval_mike.record.apply(lambda d: d['Question'])
df_eval_mike['Mike_Target_Response'] = df_eval_mike.record.apply(lambda d: d['Mike_Target_Response'])
df_eval_mike['Mike_Response'] = df_eval_mike.record.apply(lambda d: d['Mike_Response'])
df_eval_mike['cosine_mike'] = df_eval_mike.record.apply(lambda d: d['cosine_mike'])

# Extracting fields from the 'evaluation' dictionary
df_eval_mike['accuracy'] = df_eval_mike.evaluation.apply(lambda d: d['Accuracy'])
df_eval_mike['explanation'] = df_eval_mike.evaluation.apply(lambda d: d['Explanation'])

# Dropping the 'record' and 'evaluation' columns
df_eval_mike.drop(columns=['record', 'evaluation'], inplace=True)


In [121]:
# Assuming `evaluations_bella` is a list of tuples in the format (record, evaluation)
df_eval_bella = pd.DataFrame(evaluations_bella, columns=['record', 'evaluation'])

# Extracting fields from the 'record' dictionary for Bella
df_eval_bella['question'] = df_eval_bella.record.apply(lambda d: d['Question'])
df_eval_bella['Bella_Target_Response'] = df_eval_bella.record.apply(lambda d: d['Bella_Target_Response'])
df_eval_bella['Bella_Response'] = df_eval_bella.record.apply(lambda d: d['Bella_Response'])
df_eval_bella['cosine_bella'] = df_eval_bella.record.apply(lambda d: d['cosine_bella'])

# Extracting fields from the 'evaluation' dictionary
df_eval_bella['accuracy'] = df_eval_bella.evaluation.apply(lambda d: d['Accuracy'])
df_eval_bella['explanation'] = df_eval_bella.evaluation.apply(lambda d: d['Explanation'])

# Dropping the 'record' and 'evaluation' columns
df_eval_bella.drop(columns=['record', 'evaluation'], inplace=True)


In [122]:
# Assuming `evaluations_mia` is a list of tuples in the format (record, evaluation)
df_eval_mia = pd.DataFrame(evaluations_mia, columns=['record', 'evaluation'])

# Extracting fields from the 'record' dictionary for Mia
df_eval_mia['question'] = df_eval_mia.record.apply(lambda d: d['Question'])
df_eval_mia['Mia_Target_Response'] = df_eval_mia.record.apply(lambda d: d['Mia_Target_Response'])
df_eval_mia['Mia_Response'] = df_eval_mia.record.apply(lambda d: d['Mia_Response'])
df_eval_mia['cosine_mia'] = df_eval_mia.record.apply(lambda d: d['cosine_mia'])

# Extracting fields from the 'evaluation' dictionary
df_eval_mia['accuracy'] = df_eval_mia.evaluation.apply(lambda d: d['Accuracy'])
df_eval_mia['explanation'] = df_eval_mia.evaluation.apply(lambda d: d['Explanation'])

# Dropping the 'record' and 'evaluation' columns
df_eval_mia.drop(columns=['record', 'evaluation'], inplace=True)


In [123]:
df_eval_mia.head()

question  \
0                            - Скільки вам років?   
1                  - В якому місті ви народились?   
2                                – Де ви виросли?   
3                    - У якому місті ви мешкаєте?   
4  - Скільки у вас братів та сестер? Як їх звати?   

            Mia_Target_Response  \
0                            25   
1                         Львів   
2                        Сокаль   
3                         Львів   
4  Одна сестра, звати Анастасія   

                                        Mia_Response  cosine_mia    accuracy  \
0                                    Мені 25 років .    0.298563    ACCURATE   
1                             Я народилася у Львові.    0.832539    ACCURATE   
2                         Я виросла в місті Сокаль .    0.839899    ACCURATE   
3                          Я проживаю у місті Львів.    0.839686    ACCURATE   
4  У мене є одна сестра на ім'я Белла. Ми близнючки.    0.849377  INACCURATE   

                                                                                                                                                                                  explanation  
0                                                                  The generated answer provides the correct age, 25, in a grammatically correct sentence that directly answers the question.  
1                             The generated answer accurately provides the city of birth, Lviv, as the actual answer. The format is a first-person statement, but the information is correct.  
2  The generated answer is factually correct and aligns with the actual answer. It provides the name of the city where the individual grew up, matching the information in the actual answer.  
3                                                                      The generated answer is semantically equivalent to the actual answer, both stating that the city of residence is Lviv.  
4                                                     The generated answer contradicts the original answer by stating that the sister's name is Beлла, while the actual answer is Anastasiia.

In [125]:
df_eval_mike.head(3)

question  \
0            - Скільки вам років?   
1  - В якому місті ви народились?   
2                – Де ви виросли?   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   Mike_Target_Response  \
0                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    32   
1                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              місті Долина Івано-Франківської області.   
2  Загалом я виріс у місті Долина Івано-Франківської області на квартирі. В дитинстві, особливо влітку, я часто їздив в гості до бабусі та дідуся як з маминої, так і з татаової сторони. Вони проживали в селі. Мамині батьки - у селі Станківці Долинського району, а татові - у селі Нижня Луковиця Стрийського району.З маминої сторони моя бабуся з дідом завжди забороняли мені все: не міг я піти на річку чи в клуб на дискотеку. Навпаки, з тата батьки бабусі з дідом підтримували і дозволяли мені ходити на річку та на дискотеку. Тому я завжди любив відвідувати татаових батьків, де мені дозволяли робити те, що хотів.   

                                       Mike_Response  cosine_mike    accuracy  \
0                         Мені 32 роки. Як щодо вас?     0.147329    ACCURATE   
1  Я народився у місті Дрогобич. А ви де народилися?     0.815721  INACCURATE   
2         Я виріс у Дрогобичі. А де саме виросли ви?     0.846585  INACCURATE   

                                                                                                                                                                                                                                                                                                                     explanation  
0                                                                                                             The generated answer provides the correct age (32 years), which is in line with the actual answer. Additionally, it goes beyond merely providing the age by engaging in conversation, asking about the user's age.  
1                                                                                                                                     The generated answer is inaccurate as it states the evaluator was born in Drogobych, which contradicts the actual answer that they were born in the city of Dolyna, Ivan-Frankivsk region.  
2  The generated answer contradicts the actual answer as it states that the person grew up in Drogobych, 

In [126]:
df_eval_mike.to_csv('data/mike_accuracy.csv', index=False)
df_eval_mia.to_csv('data/mia_accuracy.csv', index=False)
df_eval_bella.to_csv('data/bella_accuracy.csv', index=False)
